### PR TITLE
feat: add `multiply.rs`

### PR DESCRIPTION
**Pull request by mengfansheng-git**
Given two non-negative integers num1 and num2 expressed as strings, returns the product of num1 and num2, which is also expressed as a string.
Note: You cannot use any of the built-in BigInteger libraries or directly convert the input to an integer.

### CODE REVIEW

File: src/string/multiply.rs [hunk 1]
```diff
@@ -0,0 +1,54 @@
+/*
+Given two non-negative integers num1 and num2 expressed as strings, returns the product of num1 and num2, which is also expressed as a string.
+Note: You cannot use any of the built-in BigInteger libraries or directly convert the input to an integer.
```
Comment (vil02) [inline src/string/multiply.rs:3]: This is a _problem statement_. Please change it to a proper _documentation_ (i.e. an explanation what does your code do)

File: src/string/multiply.rs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+*/
+
+pub fn multiply(num1: String, num2: String) -> String {
```
Comment (vil02) [inline src/string/multiply.rs:6]:
```suggestion
pub fn multiply(num1: &str, num2: &str) -> String {
```

File: src/string/multiply.rs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    if num1 == "0" || num2 == "0" || num1.is_empty() || num2.is_empty() {
+        return "0".to_string();
+    }
+
+    let m = num1.len();
+    let n = num2.len();
```
Comment (vil02) [inline src/string/multiply.rs:12]: What do you think about
```suggestion
    let output_size = num1.len() + num2.len();
```

File: src/string/multiply.rs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+    let mut mult = vec![0; m + n];
+
+    for (i, c1) in num1.chars().rev().enumerate() {
+        for (j, c2) in num2.chars().rev().enumerate() {
+            let n1 = c1 as u8 - b'0';
+            let n2 = c2 as u8 - b'0';
```
Comment (vil02) [inline src/string/multiply.rs:19]: What about introducing a function `to_digit`?

File: src/string/multiply.rs [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+            }
+        )*
+        }
+    }
+
+    test_multiply! {
+        multiply0: ("2".to_string(), "3".to_string(), "6"),
+        multiply1: ("123".to_string(), "456".to_string(), "56088"),
+        multiply_zero: ("0".to_string(), "222".to_string(), "0"),
+        multiply_empty: ("".to_string(), "222".to_string(), "0"),
```
Comment (vil02) [inline src/string/multiply.rs:52]:
```suggestion
        multiply0: ("2", "3", "6"),
        multiply1: ("123", "456", "56088"),
        multiply_zero: ("0", "222", "0"),
        multiply_empty: ("", "222", "0"),
        other_1: ("99", "99", "9801"),
        other_2: ("999", "99", "98901"),
        other_3: ("9999", "99", "989901"),
        other_4: ("192939", "9499596", "1832842552644"),
```

File: src/string/multiply.rs [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+#[cfg(test)]
+mod tests {
+    macro_rules! test_multiply {
+        ($($name:ident: $inputs:expr,)*) => {
+        $(
+            #[test]
+            fn $name() {
+                use super::multiply;
+                let (s, t, expected) = $inputs;
+                assert_eq!(multiply(s, t), expected);
```
Comment (vil02) [inline src/string/multiply.rs:42]:
```suggestion
                assert_eq!(multiply(s, t), expected);
                assert_eq!(multiply(t, s), expected);
```

File: src/string/multiply.rs [hunk 7]
```diff
@@ -0,0 +1,54 @@
+/*
+Given two non-negative integers num1 and num2 expressed as strings, returns the product of num1 and num2, which is also expressed as a string.
+Note: You cannot use any of the built-in BigInteger libraries or directly convert the input to an integer.
+*/
+
+pub fn multiply(num1: String, num2: String) -> String {
+    if num1 == "0" || num2 == "0" || num1.is_empty() || num2.is_empty() {
```
Comment (vil02) [inline src/string/multiply.rs:7]: I suggest to throw an error then one of the inputs is empty.

Further more your code does not handle properly the inputs with leading `0`'s and inputs with not only digits. What about negative numbers?

File: DIRECTORY.md
Comment (vil02) [inline DIRECTORY.md:1]: What do you think about moving this file into the [src/big_integer](https://github.com/TheAlgorithms/Rust/tree/master/src/big_integer)?

**Review by vil02 — CHANGES_REQUESTED** [review]

File: src/string/multiply.rs (hunk 7 above)
Comment (mengfansheng-git) [reply src/string/multiply.rs:7]: Starting with 0 and negative number strings I would like to follow up with optimizations

**Comment by mengfansheng-git** [comment]
This clippy error should be a machine issue

**Comment by vil02** [comment]
> This clippy error should be a machine issue

I think these errors are related to the recent update of the clippy. I will make a suitable fix in the evening.

**Comment by mengfansheng-git** [comment]
> > This clippy error should be a machine issue
> 
> I think these errors are related to the recent update of the clippy. I will make a suitable fix in the evening.

ok

**Comment by vil02** [comment]
When you will update/rebase your branch, the CI should pass.

File: src/big_integer/multiply.rs [hunk 8]
```diff
@@ -0,0 +1,64 @@
+/// Compute the multiplier of two strings of numbers
+///（Negative numbers and strings of numbers starting with 0 are not supported at this time）
+
```
Comment (vil02) [inline src/big_integer/multiply.rs:3]: Maybe something like
```suggestion
/// Performs a long multiplication on numbers represented by strings.
```
Also please document, that both of the strings should actually represent non-negative numbers.

File: src/big_integer/multiply.rs [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+        other_1: ("99", "99", "9801"),
+        other_2: ("999", "99", "98901"),
+        other_3: ("9999", "99", "989901"),
+        other_4: ("192939", "9499596", "1832842552644"),
+    }
+    #[test]
+    #[should_panic]
+    fn panic_when_inputs_is_empty() {
+        multiply("", "121");
+        multiply("", "");
```
Comment (vil02) [inline src/big_integer/multiply.rs:62]: This line is never executed, because the line above panics.

File: src/big_integer/multiply.rs [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    let output_size = num1.len() + num2.len();
+
+    let mut mult = vec![0; output_size];
+    for (i, c1) in num1.chars().rev().enumerate() {
+        for (j, c2) in num2.chars().rev().enumerate() {
+            let n1 = c1.to_digit(10).unwrap();
+            let n2 = c2.to_digit(10).unwrap();
+            let mul = n1 * n2;
```
Comment (vil02) [inline src/big_integer/multiply.rs:20]:
```suggestion
            let mul = c1.to_digit(10).unwrap() * c2.to_digit(10).unwrap();
```

**Review by vil02 — CHANGES_REQUESTED** [review]
`multiply("03", "2")` returns `"06"`.

File: src/big_integer/multiply.rs (hunk 9 above)
Comment (mengfansheng-git) [reply src/big_integer/multiply.rs:62]: Referring to the test in string/hamming_distance.rs，The test executes 8, it should be executed here.

**Comment by mengfansheng-git** [comment]
> Member

multiply2: ("03", "2", "06")
multiply3: ("03", "02", "006")
It currently returns. Do I need to add it to test?

**Comment by mengfansheng-git** [comment]
> `multiply("03", "2")`返回`"06"`。

multiply2: ("03", "2", "06")
multiply3: ("03", "02", "006")
It currently returns. Do I need to add it to test?

File: src/big_integer/multiply.rs (hunk 9 above)
Comment (vil02) [reply src/big_integer/multiply.rs:59]: What do you mean?

The line `multiply("", "121");` panics, so `multiply("", "");` is not executed.

**Comment by vil02** [comment]
> > `multiply("03", "2")`返回`"06"`。
> 
> multiply2: ("03", "2", "06") multiply3: ("03", "02", "006") It currently returns. Do I need to add it to test?

I would suggest to exclude inputs like `"03"`, `"003"` or `"X3"`, simply because they are not properly formatted numbers.

File: src/big_integer/multiply.rs (hunk 9 above)
Comment (mengfansheng-git) [reply src/big_integer/multiply.rs:59]: realize

**Comment by mengfansheng-git** [comment]
> > > `multiply("03", "2")`返回`"06"`。
> > 
> > 
> > multiply2: ("03", "2", "06") multiply3: ("03", "02", "006") It currently returns. Do I need to add it to test?
> 
> I would suggest to exclude inputs like `"03"`, `"003"` or `"X3"`, simply because they are not properly formatted numbers.

I'll add a string formatting judgment

File: src/big_integer/multiply.rs [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+    #[should_panic]
+    fn panic_when_inputs_is_empty() {
+        multiply("", "121");
+    }
+
+    #[test]
+    #[should_panic]
+    fn panic_when_inputs_is_abnormal() {
+        multiply("012", "1a1");
+    }
```
Comment (vil02) [inline src/big_integer/multiply.rs:72]:
```suggestion
    macro_rules! test_multiply_with_wrong_input {
        ($($name:ident: $inputs:expr,)*) => {
        $(
            #[test]
            #[should_panic]
            fn $name() {
                let (s, t) = $inputs;
                multiply(s, t);
            }
        )*
        }
    }

    test_multiply_with_wrong_input! {
        empty_input: ("", "121"),
        leading_zeros: ("001", "3"),
        wrong_characters: ("2", "12d4"),
        //wrong_input_and_zero_1: ("0", "x"),
        //wrong_input_and_zero_2: ("y", "0"),
    }
```

**Review by vil02 — CHANGES_REQUESTED** [review]

File: src/big_integer/multiply.rs [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+        }
+    }
+    if mult[output_size - 1] == 0 {
+        mult.pop();
+    }
+    mult.iter().rev().map(|&n| n.to_string()).collect()
+}
+
+pub fn is_valid_nonnegative(num: &str) -> bool {
+    !(!num.chars().all(char::is_numeric) || (&num[0..1] == "0" && num.len() > 1) || num.is_empty())
```
Comment (vil02) [inline src/big_integer/multiply.rs:30]:
```suggestion
    num.chars().all(char::is_numeric) && !num.is_empty() && (!num.starts_with('0') || num == "0")
```

File: src/big_integer/multiply.rs [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+            fn $name() {
+                let (s, t) = $inputs;
+                multiply(s, t);
+            }
+        )*
+        }
+    }
+    test_multiply_with_wrong_input! {
+        empty_input: ("", "121"),
+        leading_zeros: ("001", "3"),
```
Comment (vil02) [inline src/big_integer/multiply.rs:73]:
```suggestion
        leading_zero: ("01", "3"),
```

File: src/big_integer/multiply.rs [hunk 14]
```diff
@@ -0,0 +1,78 @@
+/// Performs long multiplication on string representations of non-negative numbers.
+
```
Comment (vil02) [inline src/big_integer/multiply.rs:2]:
```suggestion
```

**Review by vil02 — CHANGES_REQUESTED** [review]

**Review by vil02 — APPROVED** [review]

**Pull request merged**